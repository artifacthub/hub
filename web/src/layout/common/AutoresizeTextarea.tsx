import isUndefined from 'lodash/isUndefined';
import React, { useCallback, useEffect, useState } from 'react';

import styles from './AutoresizeTextarea.module.css';

interface Props {
  value?: string;
  name: string;
  disabled?: boolean;
  required?: boolean;
  invalidText?: string;
  minRows?: number;
  maxRows?: number;
  onChange?: (e: React.ChangeEvent<HTMLTextAreaElement>) => void;
}

const DEFAULT_MIN_ROWS = 3;

const AutoresizeTextarea = (props: Props) => {
  const [rows, setRows] = useState<number>(6);
  const [textValue, setTextValue] = useState<string>(props.value || '');

  const calculateHeight = useCallback(
    (value?: string) => {
      const txtValue = value || props.value || '';
      const textLength = txtValue.split('\n').length;
      const minRows = props.minRows || DEFAULT_MIN_ROWS;
      const newRowsNumber = textLength < minRows ? minRows : textLength;
      setRows(props.maxRows ? (newRowsNumber > props.maxRows ? props.maxRows : newRowsNumber) : newRowsNumber);
    },
    [props.maxRows, props.minRows, props.value]
  );

  useEffect(() => {
    setTextValue(props.value || '');
    calculateHeight();
  }, [calculateHeight, props.value]);

  return (
    <>
      <textarea
        data-testid={`${props.name}Textarea`}
        className={`form-control ${styles.textarea}`}
        rows={rows}
        id={props.name}
        name={props.name}
        value={textValue}
        disabled={props.disabled}
        required={props.required}
        onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => {
          calculateHeight(e.target.value);
          setTextValue(e.target.value);
          if (!isUndefined(props.onChange)) {
            props.onChange(e);
          }
        }}
        wrap="off"
        spellCheck="false"
      />
      {!isUndefined(props.invalidText) && <div className="invalid-feedback mt-0">{props.invalidText}</div>}
    </>
  );
};

export default AutoresizeTextarea;
