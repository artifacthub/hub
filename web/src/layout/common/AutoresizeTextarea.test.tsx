import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import AutoresizeTextarea from './AutoresizeTextarea';

interface Test {
  value: string;
  rows: number;
}

const tests: Test[] = [
  { value: 'Test', rows: 3 },
  { value: 'Line 1\nLine 2', rows: 3 },
  { value: 'Line 1\nLine 2\nLine 3', rows: 3 },
  { value: 'Line 1\nLine 2\nLine 3\nLine 4', rows: 4 },
  { value: 'Line 1\nLine 2\nLine 3\nLine 4\nLine 5', rows: 5 },
  { value: 'Line 1\nLine 2\nLine 3\nLine 4\nLine 5\nLine 6\nLine 7\nLine 8\nLine 9\nLine 10', rows: 10 },
];

const mockOnChange = jest.fn();

const defaultProps = {
  name: 'test',
  onChange: mockOnChange,
};

describe('AutoresizeTextarea', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<AutoresizeTextarea {...defaultProps} />);
    expect(asFragment).toMatchSnapshot();
  });

  for (let i = 0; i < tests.length; i++) {
    it('renders component', () => {
      const { getByTestId } = render(<AutoresizeTextarea {...defaultProps} value={tests[i].value} />);
      const textarea = getByTestId('testTextarea');
      expect(textarea).toHaveProperty('rows', tests[i].rows);
    });
  }

  it('calls onChange when textarea content is changed', () => {
    const { getByTestId } = render(<AutoresizeTextarea {...defaultProps} />);
    const textarea = getByTestId('testTextarea');
    fireEvent.change(textarea, { target: { value: 'test' } });

    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });
});
