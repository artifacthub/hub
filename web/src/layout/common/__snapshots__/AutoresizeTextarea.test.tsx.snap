// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoresizeTextarea creates snapshot 1`] = `[Function]`;
